nixos/modules/services/networking/seafile.nix

```diff
@@ -0,0 +1,272 @@
+            ${pkgs.sqlite}/bin/sqlite3 ${ccnetDir}/OrgMgr/orgmgr.db ".read ${pkgs.seafile-server}/share/seafile/sql/org.sql"
+            ${pkgs.sqlite}/bin/sqlite3 ${ccnetDir}/PeerMgr/usermgr.db ".read ${pkgs.seafile-server}/share/seafile/sql/user.sql"
+            ${pkgs.sqlite}/bin/sqlite3 ${dataDir}/seafile.db ".read ${pkgs.seafile-server}/share/seafile/sql/seafile.sql"
+        fi
+      '';
+    };
+
+    systemd.services.seahub =
+      let penv = (pkgs.python3.withPackages (ps: with ps; [ gunicorn seahub ]));
+      in {
+        description = "Seafile Server Web Frontend";
+        wantedBy = [ "seafile.target" ];
+        partOf = [ "seafile.target" ];
+        after = [ "network.target" "seaf-server.service" ];
+        requires = [ "seaf-server.service" ];
+        restartIfChanged = true;
+        restartTriggers = [ (seafileConf) ];
+        environment = {
+          PYTHONPATH =
+            "${pkgs.python3Packages.seahub}/thirdpart:${pkgs.python3Packages.seahub}:${penv}/${python.sitePackages}";
+          DJANGO_SETTINGS_MODULE = "seahub.settings";
+          CCNET_CONF_DIR = ccnetDir;
+          SEAFILE_CONF_DIR = dataDir;
+          SEAFILE_CENTRAL_CONF_DIR = centralConfigDir;
+          SEAFILE_RPC_PIPE_PATH = "/run/seafile";
+        };
+        serviceConfig = {
+          User = "seafile";
+          Group = "seafile";
+          DynamicUser = true;
+          ProtectHome = true;
+          PrivateUsers = true;
+          PrivateDevices = true;
+          ProtectClock = true;
+          ProtectHostname = true;
+          ProtectProc = "invisible";
+          ProtectKernelModules = true;
+          ProtectKernelTunables = true;
+          ProtectKernelLogs = true;
+          ProtectControlGroups = true;
+          RestrictNamespaces = true;
+          RestrictAddressFamilies = [ "AF_UNIX" ];
+          LockPersonality = true;
+          RestrictRealtime = true;
+          RestrictSUIDSGID = true;
+          MemoryDenyWriteExecute = true;
+          SystemCallArchitectures = "native";
+          RuntimeDirectory = "seahub";
+          StateDirectory = "seafile";
+          LogsDirectory = "seafile";
+          ExecStart = ''
+            ${penv}/bin/gunicorn seahub.wsgi:application \
+            --name seahub \
+            --workers ${toString cfg.workers} \
+            --log-level=info \
+            --preload \
+            --timeout=1200 \
+            --limit-request-line=8190 \
+            --bind unix:/run/seahub/gunicorn.sock
+          '';
+        };
+        preStart = ''
+          if [ ! -d "${seahubDir}" ]; then
+              mkdir ${seahubDir}
+              ln -sf ${seahubSettings} ${centralConfigDir}/seahub_settings.py
+              # link media into local seahubDir, for avatars
+              mkdir -p ${seahubDir}/media
+              for m in ${pkgs.python3Packages.seahub}/media/*/ ; do
+                ln -s $m ${seahubDir}/media/
+              done
```

Comment:
Why can't `seafile` tell where these files are in the nix store?

Reply:
We can set the `MEDIA_ROOT` in seahub settings, but seahub will attempt to upload avatars in there. I'll investigate to see if we can split the avatar directory from the other ones.

Reply:
I've tested to split the two but this is not possible, the app expects "avatars" to be a sub directory of media:
```
[ERROR] seahub.api2.endpoints.user_avatar:72 post The joined path (/var/lib/seafile/seahub/media/avatars/9/d/d64c8112b59fbad4889be2c780f3dc/2fac4ed4f235928d0e79e46ae85dbc40.jpg) is located outside of the base path component (/nix/store/q84i4j6f2g87iia39l1vpj71hz1njbq0-python3.8-seahub-8.0.4/media)
```

Reply:
Understood :+1: